RF: RIA remote

bpoldrack opened the pull request:
First of the announced follow up PR sitting on top of PR #4164 .

This includes:

 - [x] move creation routines out of the special remote itself, leading to `initremote` and `prepare` to expect an existing and accessible storage location. Furthermore split store from dataset representation in that regard, meaning two creation routines (one for a store and its layout and for a dataset within a store and its layout)
 - [x] rewrite tests
 - [x] New concept of storing rewritten URL in special remote config and its uuid in the bare repo's config in a store. `clone` gets a fallback solution in `postclone_ria` in case we cloned from a RIA store and `annex-init` didn't enable an ORA remote. We would now look for an ORA remote UUID in origin's config and try to enable this one using the `ria+` URL we cloned from. This is useful for example when the special remote URL used to publish to the store isn't valid from the point of view of a consumer. Note, that this currently supports `file:` and `ssh:` URLS. Meaningful `http:`support depends on actual support for HTTP in the special remote itself.
 - [x] renaming of the special remote and related configs, etc. from RIA to ORA. Includes command renaming `ria-export-archive` to `ora-export-archive`
- [x] exclude `@turtle` tests in AppVeyor, too. Not just in Travis.


There is still some RF'ing to do and issues like #4436 and #4437 to address/investigate. However, this should be based on the remaining RF'ing which depends on PR #4068. Plan is to build a remote execution abstraction using the approach of `SSHRemoteIO` in the special remote but with the async based new runner.

mih commented:
Needs to start from #4260 now

bpoldrack commented:
Update: rebased on master and hopefully fixed the fallout. `test_ria_basics` should now pass again.
Proceeding to rewrite/restructure the remaining tests to account for change that requires creation of the store location beforehand now.

mih commented on datalad/core/distributed/clone.py:
```diff
@@ -585,15 +585,125 @@ def postclonecfg_ria(ds, props):
+        # datalad.ora-remote.uuid as stored by create-sibling-ria and enable try
+        # enabling that one.
+        lgr.debug("Found no autoenabled ORA special remote. Trying to look it"
+                  "up in source config ...")
```
One of the lines needs a space.

bpoldrack replied:
No idea why this isn't marked outdated. But this is done.

mih commented on datalad/core/distributed/clone.py:
```diff
@@ -585,15 +585,125 @@ def postclonecfg_ria(ds, props):
+        # get that config file the same way:
+
+        config_content = None
+        scheme = props['source'].split(':')[0][4:]
```
This assumes the `props['source']` is something like `ria+...://` which is true at this point (with the current code), but it could be relatively painful to decode this fact back from this line in a future RF. Maybe add a comment? Or strip `len('ria+')`?

bpoldrack replied:
Agree. `lstrip("ria+")` might be most explicit.

mih replied:
Careful with using `*strip()`: 'aaaaa'.lstrip('a')

Better stick to your pattern, but with `len()`

bpoldrack replied:
Yeah - not concerned with "ria+ria+ria+" URLs actually, but still simply faster to with slicing.

mih replied:
It would also strip `air+` or `ira+` or `ai`. OMG it removes any AI...

bpoldrack replied:
> OMG it removes any AI...

Sounds like a feature ;-)

mih commented on datalad/core/distributed/clone.py:
```diff
@@ -585,15 +585,125 @@ def postclonecfg_ria(ds, props):
+            # HTTP:
+            lgr.debug("Trying to enable remote origin for git-annex")
+            ds.repo.enable_remote('origin')
+            ds.repo.fsck(remote='origin', fast=True)
```
I assume the last line is needed, because git annex isn't aware of any availability at "origin", because the data was copied through the special remote. If that is correct, then I think we should not do this. The result would be that annex believe that two copies of the file exist, where there is only one -- thereby invalidating any minimum redundancy requirements.

bpoldrack replied:
I'm struggling with the need for this as well. There are several possible cases: We could have store with no special remote at all, in which case `fsck` should be superfluous. So, even in case of HTTP we should prob. check origin's config and not call `fsck` if there's no special remote uuid to be found.
In case a special remote was indeed used for upload to the store, I think the `fsck` is needed, since otherwise there's no way to get any content via origin. But you're right - minimum redundancy would be torpedoed. I did not yet play with `sameas`. May be that would be a solution. So, if we decide to try to use origin we also declare it `sameas` the special remote.

mih replied:
I think, if there nothing equivalent to `sameas` then the approach must be dropped.

bpoldrack replied:
Played a bit with passing `sameas` and `uuid` parameters, but no success so far. So - will drop HTTP for now and do a proper implementation in ORA remote. But won't blow up this PR even more.

mih commented on datalad/core/distributed/clone.py:
```diff
@@ -585,15 +585,125 @@ def postclonecfg_ria(ds, props):
+            try:
+                # TODO: double-check path
+                response = op.read_file(cfg_path)
+                config_content = response
```
why keep `response`?

bpoldrack replied:
This can go indeed. Separation was an intermediate need.

mih commented on datalad/core/distributed/clone.py:
```diff
@@ -585,15 +585,125 @@ def postclonecfg_ria(ds, props):
+                config_content = response
+            except RIARemoteError as e:
+                lgr.debug("Failed to get config file from source:\n%s",
+                          exc_str(e))
```
Just a side-comment: I don't know why we have `exc_str()` (compares to having a `__str__()` in our exceptions), but there is a cost for each *not shown* debug message for using it -- in contrast to just passing `e` and let stringification happen later, if actually needed.

bpoldrack replied:
`exc_str` is about getting its original location from traceback. TBH I forgot when exactly this is needed and when it's superfluous.

yarikoptic replied:
It did come super useful to me many times when I had to dig out how the heck an exception came about to be logged (especially is swallowed upstairs somewhere).  We have `DATALAD_EXC_STR_TBLIMIT` env variable to instruct to include such information.

As for performance -- if you are hitting exceptions, I think a few `ms` spent on collecting a bit more data would not really be what causes bad UX ;)

yarikoptic replied:
<details>
<summary>Did some basic timing of exc_str. Takes ~2ms on a rudimentary use case. IMHO it is worth a "wait".  2ms means that you would wait for 1 sec if you manage to throw and catch 500 exceptions.  Did not time how long logging itself would be taking in that case etc</summary> 

```shell
$> DATALAD_EXC_STR_TBLIMIT=10 python ../trash/bench_exc_str.py
str:     0.000009 = my exception
exc_str: 0.001914 = my exception [bench_exc_str.py:<module>:12,bench_exc_str.py:f1:9,bench_exc_str.py:f2:8,bench_exc_str.py:f3:7]
(dev3) 1 22499.....................................:Tue 28 Apr 2020 12:56:58 PM EDT:.
(git)lena:~datalad/datalad-master[master]git
$> python ../trash/bench_exc_str.py        
str:     0.000008 = my exception
exc_str: 0.001510 = my exception [bench_exc_str.py:f3:7]
(dev3) 1 22500.....................................:Tue 28 Apr 2020 12:57:03 PM EDT:.
(git)lena:~datalad/datalad-master[master]git
$> cat ../trash/bench_exc_str.py
import timeit
from datalad.dochelpers import exc_str

def f1():
    def f2():
        def f3():
            raise IndexError("my exception")
        f3()
    f2()

try:
    f1()
except Exception as e:
    g = {'e': e, 'exc_str': exc_str}
    print("str:     %f = %s" % (timeit.timeit("str(e)", globals=g, number=100), str(e)))
    print("exc_str: %f = %s" % (timeit.timeit("exc_str(e)", globals=g, number=100), exc_str(e)))

```
</details>

mih commented on datalad/core/distributed/clone.py:
```diff
@@ -585,15 +585,125 @@ def postclonecfg_ria(ds, props):
+                config_content = response
+            except RIARemoteError as e:
+                lgr.debug("Failed to get config file from source:\n%s",
+                          exc_str(e))
```
Maybe I missed it, but is the `ssh` conditional completely identical to the `file` conditional? Can this duplication be avoided?

bpoldrack replied:
All that's different is the IO class to be used (which is to be replaced by the command abstraction PR). So, should be unified with the command abstraction RF. Determination of the actual path to access is also slightly different ATM, but should become identical via the new command abstractions.

bpoldrack replied:
Should be fine to finish polishing that with PR #4068

mih commented on datalad/core/distributed/clone.py:
```diff
@@ -585,15 +585,125 @@ def postclonecfg_ria(ds, props):
+                          exc_str(e))
+        else:
+            lgr.debug("Unknown URL-Scheme in %s. Can handle HTTP, SSH or FILE."
+                      "", props['source'])
```
What is the `""` for?

bpoldrack replied:
hehe. Prob. used a `format` call before. In that case I often use `"".format(...` in a new line.

mih commented on datalad/core/distributed/clone.py:
```diff
@@ -585,15 +585,125 @@ def postclonecfg_ria(ds, props):
+                    org_uuid = result['stdout'].strip()
+                except CommandError:
+                    # doesn't contain what we are looking for
+                    lgr.debug("Found nothing in source config.")
```
Maybe `Found no UUID for ORA special remote at 'origin'`?

bpoldrack replied:
Yes, better.

mih commented on datalad/core/distributed/clone.py:
```diff
@@ -585,15 +585,125 @@ def postclonecfg_ria(ds, props):
+            new_url = props['source'].split('#')[0]  # we only need the store
+            ds.repo.enable_remote(srs[org_uuid]['name'],
+                                  options=['url={}'.format(new_url)]
+                                  )
```
Should protect against `org_uuid` not being a key in `srs`. Or is this impossible? IOW can it be that the store has a remote on file, but the special remote config in the dataset is not there? I think it can happen (git-annex branch not pushed/uptodate, maybe even special remote config removed).

bpoldrack replied:
Yes, seems possible to happen.

mih commented on datalad/core/distributed/clone.py:
```diff
@@ -585,15 +585,125 @@ def postclonecfg_ria(ds, props):
+                                  )
+
     elif len(ria_remotes) == 1:
         yield from ds.siblings('configure',
```
All autoenabling happens in a conditional branch that prevents this line from being reached, so "discovered" special remotes will not be set up as publication dependencies.

bpoldrack replied:
Thx! Missed that.

kyleam commented:
@bpoldrack Just a heads up: An old run for this on appveyor is approaching the 1 hr limit (another from the same build already hit it), and there are several queued builds for old iterations of this PR.  I'm going to prune those so that appveyor isn't clogged up.

bpoldrack commented:
@kyleam : Yes, feel free to cancel those any time. I think this PR is "feature complete" now, so will try to tackle failing tests (like crippled FS) and figure what's still running forever and not marked `turtle` yet only now.

mih commented:
@bpoldrack this should be a draft PR, I think.

bpoldrack commented:
Note, that failure in CrippledFS github workflow is in `test_create_sibling_gitlab`. No Idea how this could be related.
AppVeyor currently hangs (#4436). I don't think this is reasonably addressable within this PR. If we agree on that, I'll skip the test on windows and address it in context of rewriting SSH remote execution based on command abstractions and async.

mih commented on appveyor.yml:
```diff
@@ -7,6 +7,8 @@ environment:
       MINICONDA: C:\Miniconda35
       DATALAD_TESTS_SSH: 1
+      DATALAD_LOG_LEVEL: 5
+      DATALAD_LOG_TRACEBACK: "collide"
```
Why those? Travis doesn't do that.

bpoldrack replied:
Left over. Will remove alongside skipping the hanging test on windows (if we agree on that).

mih commented on datalad/distributed/create_sibling_ria.py:
```diff
@@ -52,9 +56,11 @@
 )
 from datalad.log import log_progress
-from datalad.distributed.ria_utils import (
+from datalad.customremotes.ria_utils import (
```
I don't like this move. We should stick to the plan to slowly move code into the new package structure. https://github.com/datalad/datalad/issues/3192

bpoldrack replied:
Yeah, I'm not entirely clear about how to deal with the actual special remote. #3192 is mostly concerned with "commands". In #4164 you moved parts into `customremotes` and parts into 
`distributed`. I had the impression that the actual special remote should be in `customremotes` while related commands are supposed to go into `distributed`.

mih replied:
I can live with that. Thx for the explanation.

mih commented on datalad/distributed/ora_export_archive.py:
```diff
@@ -45,23 +45,23 @@
 @build_doc
-class RIAExportArchive(Interface):
-    """Export an archive of a local annex object store for the RIA remote.
+class ORAExportArchive(Interface):
```
I wonder, if we rename, should we rename to `export_archive_ora` or at least `export_ora_archive`?

bpoldrack replied:
Agree. To me `export_archive_ora` feels better. We are exporting *for* use by ORA, not from something used by ORA.

bpoldrack replied:
Done.

mih commented:
@bpoldrack I have no issues with disabling Windows tests and fixing the bugs later. Usage on Windows is still largely hypothetical, and we need to make a release.

mih commented on datalad/core/distributed/clone.py:
```diff
@@ -582,20 +582,136 @@ def postclonecfg_ria(ds, props):
-                   if s.get('annex-externaltype', None) == 'ora'
-    ]
+                   if s.get('annex-externaltype', None) == 'ora']
     if not ria_remotes:
```
I know that this isn't introduced here, but shouldn't the entire lookup operation be conditions upon the actual presence of _any_ ora special remote configuration, i.e. look into `AnnexRepo.get_special_remote()`. That would be cheap, but all the remote talking is relatively expensive. For example, the HCP datasets live in a RIA store but none has an ORA remote configured. It already takes 5+ hours to clone them. This would make it even longer.

bpoldrack replied:
Agree. But to be solved separately, right?

mih replied:
I can fix that in master right now. Conflicts will then have to be resolved in here. I am OK with either approach.

mih replied:
No, I take that back. The problem doesn't exist in master. It is introduced here.

mih replied:
I pushed a fix in here.

bpoldrack replied:
I thought replacing the `siblings` call right away and collect info on all special remotes (enabled or not) in one call. Anyways - your commit solves the bigger issue. Agree.

kyleam commented:
Sorry, I've canceled the remaining AppVeyor builds for this PR.  Those one hour hangs are really clogging up the queued jobs given AppVeyor runs one job at a time.  If it's just one or a few tests hanging, perhaps decorating the tests with `@nose.timed(<some reasonable value>)` would be a good way to debug these.

mih commented:
Trying this code I see this:
```
dl clone ria+ssh://bulk1.htc.inm7.de/ds/cat_12.5/srv#f2e9193e-8401-11ea-99ab-a0369f287950 myclone
[INFO   ] RIA store doesn't exist at /ds/cat_12.5/srv                                        
[INFO   ] Configure additional publication dependency on "inm7-storage"                      
configure-sibling(ok): . (sibling)
install(ok): /tmp/myclone (dataset)
action summary:
  configure-sibling (ok: 1)
  install (ok: 1)
```

The store does exist, cloning and getting work fine. The message is wrong, or misleading.

mih commented:
Performance anecdote: Rsyncing a dataset store location vs getting a full dataset (7k files, 23GB total, all files via ora-special remote) is ~70% faster. I am not sure which pieces in the (obviously more complicated and more capable) stack contribute to the slowdown, but it might be worth investigation at some point.

mih commented:
This is broken for `GitRepo`:

```
% datalad create --no-annex
% datalad create-sibling-ria -d . -s inm7 --trust-level trust ria+file:///ds/cat_12.5/srv
[INFO   ] create siblings 'inm7' and 'inm7-storage' ... 
[ERROR  ] 'GitRepo' object has no attribute 'init_remote' [create_sibling_ria.py:_create_sibling_ria:491] (AttributeError) 
```

It should not attempt to enable a storage sibling and then crash for a non-annex dataset.

mih commented:
Code seems to have issues with plain `GitRepo`:

```
% rm -r /ds/cat_12.5/srv/653/8a7fe-8ac7-11ea-83ba-a0369f287950 
rm: cannot remove '/ds/cat_12.5/srv/653/8a7fe-8ac7-11ea-83ba-a0369f287950': No such file or directory
% dl create-sibling-ria -d . -s inm7 ria+file:///ds/cat_12.5/srv --no-storage-sibling
[INFO   ] create sibling 'inm7' ... 
[ERROR  ] [Errno 17] File exists: '/ds/cat_12.5/srv/653/8a7fe-8ac7-11ea-83ba-a0369f287950' [pathlib.py:mkdir:1251] (FileExistsError)
```

bpoldrack commented:
Plain git issues addressed in https://github.com/datalad/datalad/pull/4203/commits/18e45e49586ae54b2ef484094296734be031fb09

bpoldrack commented:
For now skipping ORA tests completely. #4469
No time to investigate ATM.

bpoldrack commented:
While above mentioned call worked for @mih as well, the actual issue was still not solved.

Issue was that the `[INFO   ] RIA store doesn't exist at /ds/cat_12.5/srv ` is a factually correct but irrelevant failure. It was checking that path locally rather than at the store's machine via SSH. This is because the special remote URL committed in the dataset is a `ria+file:` URL, while we cloned on a different machine via `ria+ssh:`. Therefore autoenabling during `git-annex-init` failed (correctly!) and new feature in `datalad-clone` fix the situation by using the URL we cloned from to reconfigure the special remote.
Addressed in https://github.com/datalad/datalad/pull/4203/commits/58016f8f9ab2683ec08184d4f495a306b6613bb5

yarikoptic commented on datalad/core/distributed/clone.py:
```diff
@@ -581,21 +581,141 @@ def postclonecfg_ria(ds, props):
+    ora_remotes = [s for s in ds.siblings('query', result_renderer='disabled')
+                   if s.get('annex-externaltype', None) == 'ora']
+    if not ora_remotes and any(
+            r.get('externaltype', None) == 'ora'
```
although I keep doing it as well (but is trying to break the habit), but note that `, None` is not necessary. That is the point of `.get` vs `__getitem__` that it would return `None` if key is not present - there is a default `None`

yarikoptic commented on datalad/core/distributed/clone.py:
```diff
@@ -581,21 +581,141 @@ def postclonecfg_ria(ds, props):
+        # 3. And read it
+        org_uuid = None
+        if config_content:
+            from datalad.utils import make_tempfile
```
no reason to delay imports from core modules such as `datalad.utils` (and probably such as `.support.network`) all the way to the point of use.  Those modules would get imported anyways somewhere else in the code and should  not be taxing on performance.

yarikoptic commented on datalad/core/distributed/clone.py:
```diff
@@ -581,21 +581,141 @@ def postclonecfg_ria(ds, props):
+    if ora_remotes:
+        if len(ora_remotes) == 1:
+            yield from ds.siblings('configure',
+                                   name='origin',
```
FWIW it might be worth defining a const like `ORA_REMOTE_NAME = 'origin'` and using through the code instead of hardcoding 'origin' everywhere.  Might help later on if multiple remotes providing ORA would be supported etc.

yarikoptic commented on datalad/core/distributed/clone.py:
```diff
@@ -581,21 +581,141 @@ def postclonecfg_ria(ds, props):
+        # First figure whether we cloned via SSH, HTTP or local path and then
+        # get that config file the same way:
+        config_content = None
+        scheme = props['giturl'].split(':')[0]
```
since we are all about micro-optimizations these days,
```suggestion
        scheme = props['giturl'].split(':', 1)[0]
```

yarikoptic commented on datalad/core/distributed/clone.py:
```diff
@@ -581,21 +581,141 @@ def postclonecfg_ria(ds, props):
+            # changed with command abstractions). So we need to get that part to
+            # have a valid path to origin's config file:
+            from datalad.support.network import URL
+            from datalad.utils import PurePosixPath
```
move imports to the top of the file, see another similar comment

yarikoptic commented on datalad/core/distributed/clone.py:
```diff
@@ -581,21 +581,141 @@ def postclonecfg_ria(ds, props):
+            # have a valid path to origin's config file:
+            from datalad.support.network import URL
+            from datalad.utils import PurePosixPath
+            cfg_path = PurePosixPath(URL(props['giturl']).path) / 'config'
```
it might be good to program a bit more defensively here and try/except for KeyError in case `giturl` was not defined, and spit out a specific meaningful exception describing what is missing `giturl` which is required

bpoldrack replied:
What would be a useful description here? It's the URL we just cloned from. If it's not in `props` we have a bug in `clone_dataset`. I wouldn't know a better message for a user. They can't do anything about it. And all we need to know to fix it is that KeyError.

yarikoptic commented on datalad/core/distributed/clone.py:
```diff
@@ -581,21 +581,141 @@ def postclonecfg_ria(ds, props):
+            try:
+                config_content = op.read_file(cfg_path)
+            except (RIARemoteError, OSError) as e:
+                lgr.debug("Failed to get config file from source:\n%s",
```
why here specifically to bother with `\n` to separate out the exception but not in other places?

yarikoptic commented on datalad/core/distributed/clone.py:
```diff
@@ -581,21 +581,141 @@ def postclonecfg_ria(ds, props):
+                GitRunner,
+                WitlessRunner,
+                StdOutCapture
+            )
```
I don't think delaying these imports also would make sense, since we are guaranteed to import those early to run `git config` etc anyways

yarikoptic commented on datalad/core/distributed/clone.py:
```diff
@@ -581,21 +581,141 @@ def postclonecfg_ria(ds, props):
+                except CommandError:
+                    # doesn't contain what we are looking for
+                    lgr.debug("Found no UUID for ORA special remote at "
+                              "'origin'")
```
worth providing details of the `CommandError` exception here in debug message.  Could crash for a variety of reasons (formatting, etc) and hiding it behind a too generic message would require more digging to figure out what went wrong

yarikoptic commented on datalad/core/distributed/clone.py:
```diff
@@ -581,21 +581,141 @@ def postclonecfg_ria(ds, props):
+                                   result_renderer='disabled')
+        else:
+            lgr.warning("Found multiple ORA remotes. Couldn't decide which "
+                        "publishing to origin should depend on: %s",
```
and what should the user do when is warning displayed - should it may be suggest something ? (remove one?)

yarikoptic commented on datalad/customremotes/ria_utils.py:
```diff
@@ -0,0 +1,158 @@
+    pass
+
+
+known_versions_objt = ['1', '2']
```
what is `_objt`.  having comments describing those constants (are they? if so - make `tuple` ?) would help new to the code developers to orient themselves

bpoldrack replied:
Added comments to the constants, but made the `tuple` suggestion a TODO only. This would require a bit more RF'ing that shouldn't delay merge right now, I think. But: Certainly useful approach. However, may be better elsewhere (define versions as `(label, description)` and reference here by `label` only) - not sure yet.

yarikoptic commented on datalad/customremotes/ria_utils.py:
```diff
@@ -0,0 +1,158 @@
+        dsobj_dir = dsgit_dir / 'annex' / 'objects'
+        return dsgit_dir, archive_dir, dsobj_dir
+    else:
+        raise ValueError("Unknown layout version: {}".format(version))
```
include information on which are known? (known_versions_dst? known_versions_objt?)

yarikoptic commented on datalad/customremotes/ria_utils.py:
```diff
@@ -0,0 +1,158 @@
+      (host, base-path, rewritten url)
+    """
+    from datalad.config import rewrite_url
+    from datalad.support.network import URL
```
no need to delay the imports... since so prevalent -- is there a shift in coding style to delay all imports until point of use we should all discuss/adopt???

yarikoptic commented on datalad/customremotes/ria_utils.py:
```diff
@@ -0,0 +1,158 @@
+            "Unexpected fragment in RIA-store URL: %s" % url_ri.fragment)
+    protocol = url_ri.scheme[4:]
+    if protocol not in ['ssh', 'file']:
+        raise ValueError("Unsupported protocol: %s" % protocol)
```
include the list of supported ones (define a const with them?)

yarikoptic commented on datalad/customremotes/ria_utils.py:
```diff
@@ -0,0 +1,158 @@
+    return url_ri.hostname if protocol == 'ssh' else None, url_ri.path, url
+
+
+def create_store(io, base_path, version):
```
Docstring would be of value here since it is not private method, and `io` is not even easy to figure out without knowing details of the codebase around here.

yarikoptic commented on datalad/customremotes/ria_utils.py:
```diff
@@ -0,0 +1,158 @@
+
+    # At store level the only version we know as of now is 1.
+    if version not in known_versions_dst:
+        raise UnknownLayoutVersion("store layout version unknown: {}"
```
and which ones are "known"? BTW, is known here (see my comment about adding a comment) is really "known" or "supported"?

yarikoptic commented on datalad/customremotes/ria_utils.py:
```diff
@@ -0,0 +1,158 @@
+            # record.
+            # Note, that a config flag after pipe symbol is fine.
+            raise ValueError("Conflicting version found at target: {}"
+                             "".format(existing_version))
```
I think it would be as fine with
```suggestion
                             .format(existing_version))
```

yarikoptic commented on datalad/customremotes/ria_utils.py:
```diff
@@ -0,0 +1,158 @@
+    io.write_file(version_file, version)
+
+
+def create_ds_in_store(io, base_path, dsid, obj_version, store_version):
```
docstring

yarikoptic commented on datalad/distributed/tests/test_ria_basics.py:
```diff
@@ -206,9 +359,133 @@ def test_version_check(path, objtree):
+    _test_gitannex(None)
+
+
+@skip_if_on_windows  # see gh-4469
```
should it ideally be `@known_failure_windows` (here and others)?

bpoldrack commented:
@yarikoptic : 
See https://github.com/datalad/datalad/pull/4203#discussion_r419226947
Apart from that addressed your comments in https://github.com/datalad/datalad/pull/4203/commits/2d3edaae007f90ed219d3ebd41b28da518f1d0a1
